pkg/resources/digitalocean/dns/dns.go

```diff
@@ -0,0 +1,501 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/digitalocean/godo"
+	"github.com/digitalocean/godo/context"
+
+	"k8s.io/kubernetes/federation/pkg/dnsprovider"
+	"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
+)
+
+const ipPlaceholder = "203.0.113.123"
+
+// DNS implements dnsprovider.Interface
+type DNS struct {
+	client *godo.Client
+}
+
+// NewProvider returns an implementation of dnsprovider.Interface
+func NewProvider(client *godo.Client) dnsprovider.Interface {
+	return &DNS{client: client}
+}
+
+// Zones returns an implementation of dnsprovider.Zones
+func (d *DNS) Zones() (dnsprovider.Zones, bool) {
+	return &zones{
+		client: d.client,
+	}, true
+}
+
+// zones is an implementation of dnsprovider.Zones
+type zones struct {
+	client *godo.Client
+}
+
+// List returns a list of all dns zones
+func (z *zones) List() ([]dnsprovider.Zone, error) {
+	domains, err := listDomains(z.client)
+	if err != nil {
+		return nil, err
+	}
+
+	var newZone *zone
+	var zones []dnsprovider.Zone
+	for _, domain := range domains {
+		newZone = &zone{
+			name:   domain.Name,
+			client: z.client,
+		}
+		zones = append(zones, newZone)
+	}
+
+	return zones, nil
+}
+
+// Add adds a new DNS zone
+func (z *zones) Add(newZone dnsprovider.Zone) (dnsprovider.Zone, error) {
+	domainCreateRequest := &godo.DomainCreateRequest{
+		Name:      newZone.Name(),
+		IPAddress: ipPlaceholder,
+	}
+
+	domain, err := createDomain(z.client, domainCreateRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	return &zone{
+		name:   domain.Name,
+		client: z.client,
+	}, nil
+}
+
+// Remove deletes a zone
+func (z *zones) Remove(zone dnsprovider.Zone) error {
+	return deleteDomain(z.client, zone.Name())
+}
+
+// New returns a new implementation of dnsprovider.Zone
+func (z *zones) New(name string) (dnsprovider.Zone, error) {
+	return &zone{
+		name:   name,
+		client: z.client,
+	}, nil
+
+}
+
+// zone implements dnsprovider.Zone
+type zone struct {
+	name   string
+	client *godo.Client
+}
+
+// Name returns the Name of a dns zone
+func (z *zone) Name() string {
+	return z.name
+}
+
+// ID returns the name of a dns zone, in DO the ID is the name
+func (z *zone) ID() string {
+	return z.name
+}
+
+// ResourceRecordSet returns an implementation of dnsprovider.ResourceRecordSets
+func (z *zone) ResourceRecordSets() (dnsprovider.ResourceRecordSets, bool) {
+	return &resourceRecordSets{zone: z, client: z.client}, true
+}
+
+// resourceRecordSets implements dnsprovider.ResourceRecordSet
+type resourceRecordSets struct {
+	zone   *zone
+	client *godo.Client
+}
+
+// List returns a list of dnsprovider.ResourceRecordSet
+func (r *resourceRecordSets) List() ([]dnsprovider.ResourceRecordSet, error) {
+	records, err := getRecords(r.client, r.zone.Name())
```

Comment:
There's no way to filter records iirc, but for the case where there's a lot of records, you can paginate. This is definitely something I'd want to address before DO support goes beta, for now I think this will be okay. (I'll leave a note though)